design/metastore.md

```diff
@@ -0,0 +1,102 @@
+One of the discover items is to select the best way we can model the information in order to have a better way to diff/merge changes in metastore data.  The key format we will use will effect how the underlaying ranges will be split, the cost of lookup of nested information and etc.
+In order to keep a reference to a state of the above model, the data we will keep depends on they we store.
+For example in case we like to store database information in our key/value storage:
+
+```
+db/db1 -> {name, description, locationUri, parameters, privileges, ownerName, ownerType}
+db/db2 -> {name, description, locationUri, parameters, privileges, ownerName, ownerType}
+...
+```
+
+Different keys holds different value format
+
+```
+db/db1/table/tbl1 -> {tableName,dbName,owner,createTime,lastAccessTime,retention,sd,partitionKeys,...}
+db/db1/table/tbl2 -> {tableName,dbName,owner,createTime,lastAccessTime,retention,sd,partitionKeys,...}
+db/db2/table/tbl1 -> {tableName,dbName,owner,createTime,lastAccessTime,retention,sd,partitionKeys,...}
+...
+```
+
+The data model that Hive Metastore provides enables more than the above. We will implement the missing entities as required by the data tools.
+
+
+### Thrift API
+
+Implementing the Hive Metastore 2.3.x will be the starting point of lakeFS's Metastore.
+Spark works with Hive Metastore version 2.3.9 (https://spark.apache.org/docs/latest/sql-data-sources-hive-tables.html), but can also work with version 3.x.
+Trino can work with Hive 2 and 3, while having specific capabilities enabled for version 3.
+Must of the Hive Metastore Thrift interface kept backward compatibility v2.3.x and v3.x interfaces are almost identical, where version 3 adds Catalog and Workload management metadata.
+
+Code generation should be considered. Having generated bride that will transfer each Thrift API call to our OpenAPI. And/or generate code for each entity described in the Thrift interface can reduce the code in order serialize, map and filter the data model.
+
+Thrift API communicate over a socket, the first step will enable non-secure communication over an additional port. A secure communication can be handled by a external TCP load-balancer first, later the proxy can supply this capability out of the box.
+
+
+Hooks - The Thrift interface include a set of API ends with '_environment_context', which enables the caller passing additional information to the hooks. The hooks mechanism in Hive Metastore is a Java interface implementation that the metastore load and calls on specific operations. We will not support this mechanism at first, but we can have a reference implementation of what to expect in case of adding it to our hooks mechanism.
```

Comment:
```suggestion
Hooks - The Thrift interface includes a set of API endpoints with the suffix '_environment_context', which allow the caller to pass additional information to the hooks. The hooks mechanism in Hive Metastore is a Java interface implementation that the metastore loads and calls on specific operations. We will not support this mechanism at first, but we can have a reference implementation of what to expect in case of adding it to our hooks mechanism.
```